### ISSUE TITLE
Win8.x Store targeted apps are rejected due to Win32 crypto APIs

### ISSUE THREAD

**Issue by hetadesu** [AArnott/PCLCrypto#97]
Due to use of bcrypt.dll

**Comment by AArnott** [comment]
You'll have to be more specific. bcrypt.dll _is_ Windows Store compliant. The MSDN documentation states that most of its functions are available for Store apps, for example [BCryptDecrypt](https://msdn.microsoft.com/en-us/library/windows/desktop/aa375391.aspx).

**Comment by hetadesu** [comment]
Supported API test

FAILED
    Supported APIs

    Error Found: The supported APIs test detected the following errors:
        API BCryptCloseAlgorithmProvider in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptCreateHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptDecrypt in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptDeriveKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptDestroyHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptDestroyKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptDestroySecret in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptEncrypt in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptEnumAlgorithms in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptExportKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptFinalizeKeyPair in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptFinishHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptFreeBuffer in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptGenRandom in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptGenerateKeyPair in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptGenerateSymmetricKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptGetProperty in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptHashData in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptImportKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptImportKeyPair in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptOpenAlgorithmProvider in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptSecretAgreement in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptSetProperty in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptSignHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API BCryptVerifySignature in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
        API FindFirstFileExA in api-ms-win-core-file-l1-2-0.dll is not supported for this application type. PInvoke.Kernel32.dll calls this API.
        API WaitForSingleObject in api-ms-win-core-synch-l1-2-0.dll is not supported for this application type. PInvoke.Kernel32.dll calls this API.
        API CancelIoEx in api-ms-win-core-io-l1-1-1.dll is not supported for this application type. PInvoke.Kernel32.dll calls this API.
        API NCryptCreatePersistedKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptDecrypt in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptDeleteKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptDeriveKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptEncrypt in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptEnumAlgorithms in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptEnumKeys in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptExportKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptFinalizeKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptFreeBuffer in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptFreeObject in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptGetProperty in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptImportKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptIsAlgSupported in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptKeyDerivation in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptOpenKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptOpenStorageProvider in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptSecretAgreement in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptSetProperty in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptSignHash in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
        API NCryptVerifySignature in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    Impact if not fixed: Using an API that is not part of the Windows SDK for Windows Store apps violates the Windows Store certification requirements.
    How to fix: Review the error messages to identify the API that is not part of the Windows SDK for Windows Store apps. Please note, apps that are built in a debug configuration or without .NET Native enabled (where applicable) can fail this test as these environments may pull in unsupported APIs. Retest your app in a release configuration, and with .NET Native enabled if applicable. See the link below for more information:
    Alternatives to Windows APIs in Windows Store apps.

**Comment by AArnott** [comment]
Thanks. What type of application are you building? Win8.0, Win8.1, UWP?
The MSDN docs clearly state that many of these are available to Store apps, which tells me there's either a tool or a doc bug somewhere.

**Comment by hetadesu** [comment]
Oh sorry I forgot to mention that.

It's a Windows 8.1 application.

The best way to check if it's compatible is to run the App Cert tools locally with the dll referenced in the Win 8.1 program.

**Comment by AArnott** [comment]
Thanks. I wonder if the MSDN docs are imprecise and this is allowed in UWP but not Win8.x. I'll do some digging.

**Comment by AArnott** [comment]
BTW, in the meantime you can use an older (1.x) version of PCLCrypto, which doesn't depend on BCrypt.

**Comment by hetadesu** [comment]
Yes that is already done thanks

I detected the bcrypt problem after upgrading to 2.x. So Irolled back.



Envoyé depuis mon smartphone Samsung Galaxy.


-------- Message d'origine --------
De : Andrew Arnott <notifications@github.com>
Date : 14/05/2016 21:48 (GMT+01:00)
À : AArnott/PCLCrypto <PCLCrypto@noreply.github.com>
Cc : Grégory Douguet <gdouguet@wess-soft.com>, Author <author@noreply.github.com>
Objet : Re: [AArnott/PCLCrypto] Last version not Windows Store compliant (#97)


BTW, in the meantime you can use an older (1.x) version of PCLCrypto, which doesn't depend on BCrypt.

-

**Comment by AArnott** [comment]
In trying to repro it, I didn't get any errors about BCrypt. But I did see some PInvoke.Kernel32 errors. This is inconsistent with MSDN docs. But although it failed when certifying the Debug package, it passed when certifying the Release package. Did you try with debug or release?

**Comment by AArnott** [comment]
Oh, interesting. So for Windows 8.1 apps all the NCrypt functions were rejected. For UWP they are allowed.
So I guess PCLCrypto 2.x only works for UWP apps. I should probably update the nupkg to indicate this.

**Comment by hetadesu** [comment]
Ah yes interesting indeed.

I will think about upgrading my project to uwp.

Thanks for the enlightenment


Envoyé depuis mon smartphone Samsung Galaxy.


-------- Message d'origine --------
De : Andrew Arnott <notifications@github.com>
Date : 14/05/2016 22:16 (GMT+01:00)
À : AArnott/PCLCrypto <PCLCrypto@noreply.github.com>
Cc : Grégory Douguet <gdouguet@wess-soft.com>, Author <author@noreply.github.com>
Objet : Re: [AArnott/PCLCrypto] Last version not Windows Store compliant (#97)


Oh, interesting. So for Windows 8.1 apps all the NCrypt functions were rejected. For UWP they are allowed.
So I guess PCLCrypto 2.x only works for UWP apps. I should probably update the nupkg to indicate this.

-

**Comment by arivoir** [comment]
I had the same trouble with a Windows Phone 8.1 app.

**Comment by DenisSkygate** [comment]
Hello,
I have also some issues with PCLCrypto.
Validation and PInvoke are installed.
It's a Windows Phone 8.1 app, with a Xamarin PCL .

`{Windows.UI.Xaml.UnhandledExceptionEventArgs}
    Exception: {"This is a reference assembly and does not contain implementation. Be sure to install the PCLCrypto package into your application so the platform implementation assembly will be used at runtime."}
    Handled: false
    Message: "This is a reference assembly and does not contain implementation. Be sure to install the PCLCrypto package into your application so the platform implementation assembly will be used at runtime."
    Native View: To inspect the native object, enable native code debugging.`

**Comment by hetadesu** [comment]
Hi,
You can rollback to an earlier version which pass validation.
++

De : DenisSkygate [mailto:notifications@github.com]
Envoyé : jeudi 14 juillet 2016 18:18
À : AArnott/PCLCrypto <PCLCrypto@noreply.github.com>
Cc : Grégory Douguet <gdouguet@wess-soft.com>; Author <author@noreply.github.com>
Objet : Re: [AArnott/PCLCrypto] Win8.x Store targeted apps are rejected due to Win32 crypto APIs (#97)


Hello,
I have also some issues with PCLCrypto.
Validation and PInvoke are installed.
It's a Windows Phone 8.1 app, with a Xamarin PCL .

{Windows.UI.Xaml.UnhandledExceptionEventArgs}
Exception: {"This is a reference assembly and does not contain implementation. Be sure to install the PCLCrypto package into your application so the platform implementation assembly will be used at runtime."}
Handled: false
Message: "This is a reference assembly and does not contain implementation. Be sure to install the PCLCrypto package into your application so the platform implementation assembly will be used at runtime."
Native View: To inspect the native object, enable native code debugging.

**Comment by dvsekhvalnov** [comment]
Hi @AArnott , any ETA for the fix if possible?

Thank you.

**Comment by AArnott** [comment]
I wish I had one. The only "fix" I know of right now is to target UWP apps. Microsoft has acknowledged the bug on their side, but I suspect their motivation to fix the process for 8.x targeted apps is relatively low.

**Comment by dvsekhvalnov** [comment]
Got it. Thank you.

**Comment by Magic73** [comment]
I had the same issue with UWP application, but with ncrypt.dll.

API BCryptCloseAlgorithmProvider in ncrypt.dll is not supported for this application type. App.dll calls this API.

I'm using PCLCrypto 2.0.147, with Xamarin.

**Comment by dtaylorus** [comment]
@AArnott Would you consider reverting to the 1.x implemention without the BCrypt dependency?

It isn't very useful to have a library that doesn't pass the Windows Store certification tests (at least not to me).

I spent a bunch of time building a working app and now it looks like I need to either throw all the crypto code away and start over, or use an old version of this library which I assume has bugs that have been fixed in later versions.

Also, note that targeting UWP does not pass certification, as my app is a Windows 8.1 UWP app and it fails certification to the Windows Store as of today. It's quite easy to repro the failure by running the certification test after building a Windows 8.1 UWP package for the Windows Store.

Thanks!

**Comment by AArnott** [comment]
> use an old version of this library which I assume has bugs that have been fixed in later versions.

There haven't been any security bugs in this library thus far. If an older version works for you, go for it. Mostly each new version adds features and platforms. The only real "bugs" have been in serialized key pair interop across platforms, which very few folks need.

I'm told by the Windows folks that they've updated the Store certification to allow these functions. Your local certification test may fail, but the online one should pass. Can you confirm or deny this? Even if that works, it's still feedback I'll pass onto the Windows SDK team to get an update to the local certification tool.

**Comment by dtaylorus** [comment]
@AArnott I can confirm that the online validation fails. I tried once with the default nuget installation, which installs outdated versions of all of the PInvoke dependencies, and again after updating all of the PInvoke dependencies. Online validation failed both times, exactly as it did locally.

I appreciate the workaround, but I'm sure anyone else down the road would also appreciate not having to find out the hard way that their Windows Store submission fails and on top of that, find this thread with the workaround buried somewhere above.

For anyone searching for how to get a valid Windows 8.1 build, you have to uninstall all of the nuget 2.x packages (PCLCrypto, PInvoke.*) in the right order (just keep trying to uninstall them one at a time until you eventually get rid of them all), and then install version 1.0.2.15130 (I'm not sure, but it seems like the 1.0.8x builds might also depend on the PInvoke packages). That worked for me and passed both local and online store validation for Windows 8.1. I haven't tried Windows Phone 8.1 yet.

**Comment by AArnott** [comment]
Thanks for the info. I'll dig up the old thread I had with the Windows SDK folks and hopefully can press on them to fix this.
In the short term, I think what we can do is update PInvoke to exclude the banned functions from a uwp-targeted build. That should allow PCLCrypto to update its dependencies to the latest versions and unblock you. I agree: it's useless to have Win8/UWP support in PCLCrypto if ultimately it fails to submit to the store, and can lead to lost time investing in something that will ultimately fail. Let's get this working.

Can you do me a favor and send me the failure report you're getting? Both the online version and the local test results, if possible? I'll use it both to update PInvoke to workaround the issue and to make the case to the WinSDK folks that the problem is not fixed.

**Comment by dtaylorus** [comment]
Thanks, Andrew! I've sent the failures to the email address in your profile, and here's a dump of the API failures from the online failure report:

    API NCryptCreatePersistedKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptDecrypt in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptDeleteKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptDeriveKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptEncrypt in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptEnumAlgorithms in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptEnumKeys in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptExportKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptFinalizeKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptFreeBuffer in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptFreeObject in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptGetProperty in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptImportKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptIsAlgSupported in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptKeyDerivation in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptOpenKey in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptOpenStorageProvider in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptSecretAgreement in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptSetProperty in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptSignHash in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API NCryptVerifySignature in ncrypt.dll is not supported for this application type. PInvoke.NCrypt.dll calls this API.
    API FindFirstFileExA in api-ms-win-core-file-l1-2-0.dll is not supported for this application type. PInvoke.Kernel32.dll calls this API.
    API WaitForSingleObject in api-ms-win-core-synch-l1-2-0.dll is not supported for this application type. PInvoke.Kernel32.dll calls this API.
    API CancelIoEx in api-ms-win-core-io-l1-1-1.dll is not supported for this application type. PInvoke.Kernel32.dll calls this API.
    API BCryptCloseAlgorithmProvider in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptCreateHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptDecrypt in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptDeriveKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptDestroyHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptDestroyKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptDestroySecret in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptEncrypt in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptEnumAlgorithms in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptExportKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptFinalizeKeyPair in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptFinishHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptFreeBuffer in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptGenRandom in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptGenerateKeyPair in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptGenerateSymmetricKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptGetProperty in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptHashData in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptImportKey in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptImportKeyPair in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptOpenAlgorithmProvider in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptSecretAgreement in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptSetProperty in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptSignHash in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.
    API BCryptVerifySignature in bcrypt.dll is not supported for this application type. PInvoke.BCrypt.dll calls this API.

**Comment by AArnott** [comment]
When I create a UWP app and install PCLCrypto to it, then package it up and run the cert check, it passes. I wonder if this problem is exclusive to Win8 targeted apps.

**Comment by dtaylorus** [comment]
It could be specific to Windows 8. The Windows 10 version of my app doesn't use PCLCrypto because all of the crypto API's I needed are in the .NET Standard SDK.

**Comment by AArnott** [comment]
Hmmm... When I actually add code to call PCLCrypto (such that the .NET Native linker would keep it around) the cert check fails with a vague message about disallowed APIs, but doesn't list any. This is for my UWP test app.
Getting the Windows team to fix this for Win8 would be a really tough sell.

**Comment by AArnott** [comment]
Interestingly the same test but on a Win8.1 targeted app, with PCLCrypto and code to call into it, passed when I ran the cert check.

**Comment by AArnott** [comment]
One possibility, @dtaylorus, is that your Windows SDK is old. If you can install the latest one, it might help. Although that might explain how to get it to pass locally, I'd be surprised if that resolved the cloud cert check issue. But maybe... the WinSDK is certainly responsible for how it builds locally.

**Comment by AArnott** [comment]
> When I actually add code to call PCLCrypto (such that the .NET Native linker would keep it around) the cert check fails with a vague message about disallowed APIs, but doesn't list any. This is for my UWP test app.

Red herring. Running it again made it disappear.

So I can't repro this at all with the Windows 10 RS2 SDK. That same SDK is what tests Win8.1 targeted apps, it would seem.

**Comment by dtaylorus** [comment]
The only reason I use PCLCrypto is Windows 8.

These days, .NET Standard has a reasonable set of crypto API's that work on Xamarin iOS and Android and Windows 10.

So I have no need for PCLCrypto at all on Windows 10.

However, the reality is that somewhere around half of my audience is still using Windows 8, and it looks like it will be a long time before Windows 8 becomes irrelevant.

As I'm sure you know, Windows 8 is missing some .NET crypto API's that are non-trivial to implement, namely CryptoStream, which is where PCLCrypto comes to the rescue.

It was relatively straightforward for me to use PCLCrypto on Windows 8 to implement the .NET Standard API's I use on other platforms, including Windows 10.

So for me, it's unfortunate that the current version of PCLCrypto cannot pass app store certification in the only scenario where it's necessary and relevant today--Windows 8.

**Comment by dtaylorus** [comment]
All of that to say that while it's nice that you validated PCLCrypto on Windows 10, it's not a relevant test case for me.

In my case, the relevant test case is building a Windows 8.1 app that uses PCLCrypto and actually passing the live app store certification test. (I can confirm that it does not work with the latest version of PCLCrypto.)

I understand that you may have other scenarios in mind and that mine may not be that interesting for you to support, but there it is for you to decide.

Thanks, and I appreciate the workaround! My Windows 8.1 app is live in the store using an old version of PCLCrypto, and this would have been a lot more trouble without PCLCrypto.

-David

**Comment by AArnott** [comment]
Thanks, @dtaylorus. As I mentioned, I was able to build a Win8.1 targeted app that at least passed local cert checks using Windows 10 and the latest Win10 SDK. I had asked whether you had the latest WinSDK installed in case that made the difference with how the package was built. I didn't get a response to that. 

> As I'm sure you know, Windows 8 is missing some .NET crypto API's that are non-trivial to implement, namely CryptoStream, which is where PCLCrypto comes to the rescue.

CryptoStream was actually one of the drivers to update PCLCrypto to use PInvoke. The older version of PCLCrypto that works for you has a CryptoStream that's basically faked up: it buffers everything and then encrypts it as a single operation at the end using the WinRT API. It's only the latest versions based on PInvoke that truly stream the cryptographic operation that you send through it. That's one big reason why I want to fix PCLCrypto to work for your Win8.1 app while still calling the native NCrypt/BCrypt libraries.

> Thanks, and I appreciate the workaround! My Windows 8.1 app is live in the store using an old version of PCLCrypto, and this would have been a lot more trouble without PCLCrypto.

Glad to hear it. :)

Here's what I'll do though at least till I hear more from you: I'll go ahead and submit my Win8.1 targeted app to the store and see what online certification does with it. If it works, then I'll be pretty convinced that you can unblock yourself by developing with Win10 and the latest Win10 SDK (even though you're targeting a Win8.1 app).

**Comment by arun2106** [comment]
I tried to certify with windows sdk 10, still same problem. I have tried to remove Pinvoke.Ncrypt, Bcrypt. I am getting application failure, as from pclcrypto to generate keypair, it is referring to ncrypt and failing.

**Comment by AArnott** [comment]
What version of the Windows SDK do you have installed?

**Comment by arun2106** [comment]
I have tried in Microsoft Windows 10 Enterprise (10.0.10586.0). When i upload to store, it failed there as well.

**Comment by AArnott** [comment]
I'm not talking about the build of Windows. I'm asking what version of the Windows *SDK* you have installed. This is the one I'm using:
![image](https://cloud.githubusercontent.com/assets/3548/26758214/8c056924-488c-11e7-9b39-2b17f06592df.png)

**Comment by arun2106** [comment]
![capture](https://cloud.githubusercontent.com/assets/16820055/26780782/383eaab4-49e3-11e7-8e2a-3ae244078526.PNG)

I am using Visual Studio 2015 Update 2. SDK seems to be 10.0.10240. Please see attached

**Comment by AArnott** [comment]
I believe the fix is only available in the newer SDK and perhaps the newer VS. Can you give that a try?

**Comment by arun2106** [comment]
I have tried with Visual studio 2015 with Update 3 (Windows 10 SDK included). But, in order to compile my code, it requires Window 8.1 SDKs and seems to be building with that rather than Windows 10 SDKs. Any workaround?

**Comment by AArnott** [comment]
Not sure. Can you try building with VS 2017?

**Comment by HelenMamalaki** [comment]
Rolling back to v1.x of PCL Crypto fixes the problem but Xamarin.Auth is dependent on PCLCrypto (>= 2.0.147). So I currently have an Xamarin.Windows app that I cannot publish to the store. Great.

**Comment by AArnott** [comment]
@HelenMamalaki please review the comments I've made on this issue. I do believe there is a "fix" if you'll use the latest VS and latest Windows SDK.

**Comment by dtaylorus** [comment]
@AArnott It's not possible to build a Windows 8.1 app with the Windows 10 SDK, or with VS 2017 at all. When I try to open a Windows 8.1 project with VS 2017, it says it's unsupported.

This limitation is confirmed in [the documentation](https://docs.microsoft.com/en-us/visualstudio/porting/port-migrate-and-upgrade-visual-studio-projects) where it says:

> Projects for Windows Store 8.1 and 8.0, and Windows Phone 8.1 and 8.0 are not supported in Visual Studio 2017. To maintain these apps, continue to use Visual Studio 2015.

So, to answer your question, there is no way to build a Windows 8.1 project with VS 2017 or the latest Windows 10 SDK, which means your suggested fix is not possible.

If you somehow built a Windows 8.1 app with PCLCrypto 2.x that can be submitted to the store, please let me know how you did it.

(Note that in my case Windows 10 obsoletes PCLCrypto, so I don't have any need for or interest in Windows 10 + PCLCrypto.)

Thanks!

**Comment by AArnott** [comment]
> So, to answer your question, there is no way to build a Windows 8.1 project with VS 2017

Hmm... I'm seeing the same thing now as well. I'm not sure what I saw before.

> If you somehow built a Windows 8.1 app with PCLCrypto 2.x that can be submitted to the store, please let me know how you did it.

I did it by using the latest Windows 8 SDK available. But I'll try again today and submit to the store to verify again.

**Comment by HelenMamalaki** [comment]
@AArnott Did you have any luck submitting a Windows Phone 8.1 app that uses PCLCrypto2.x app to the store?

**Comment by AArnott** [comment]
I'm afraid I haven't finished the test yet.

**Comment by kinex** [comment]
My UWP app is now rejected from Windows Store with error:

`Error Found: The supported APIs test detected the following errors:
API FindFirstFileEx in api-ms-win-core-file-l1-2-0.dll is not supported for this application type. PInvoke.Kernel32.dll calls this API.`

I am quite sure this is related to PCLCrypto because it is only package in my project referencing PInvoke.Kernel32.dll.

It is also strange that the same app passed the certification one week ago (this new update has mainly translation updates only). If I run Windows App Cert Kit locally, it is PASSED normally. I have used PCLCrypto in my app already several months and there has not been any issues in certification earlier (and I have not modified related code or updated packages recently).

What would be the best workaround?

**Comment by AArnott** [comment]
Please give *this message* a thumbs up if you target wpa81 and want this issue fixed.

**Comment by AArnott** [comment]
Please give *this message* a thumbs up if you target win8.x and want this issue fixed.

**Comment by AArnott** [comment]
lease give *this message* a thumbs up if you target uwp (i.e. Win10) and want this issue fixed.

**Comment by AArnott** [comment]
I'm in conversations with Windows Store folks now to try to get this resolved. Please up-vote any of the immediately above messages I just posted so we can tally up interest in each platform that may need fixes.

**Comment by dtaylorus** [comment]
What is wpa8.1?

I target Windows 8.1 (Store not Desktop) and Windows Phone 8.1 and would like PCLCrypto support on those two platforms.

I also target Windows 10 UWP, MacOS, iOS, and Android but do not use PCLCrypto for those platforms.

Thanks for following up!

**Comment by AArnott** [comment]
@dtaylorus please vote up the wpa81 and win8.x messages then.

**Comment by ptorr-msft** [comment]
The **technical problem** here is that `FindFirstFileEx` isn't on the allow-list because *that API doesn't exist in Windows*. It would be like trying to call an API `ILovePuppiesAndDonuts` - WACK will fail because it's an unknown API.

The actual APIs in Windows are named `FindFirstFileExA` and `FindFirstFileExW`, and .NET picks the appropriate one (typically the `W` version unless you specifically ask for `A`). 

That's the reason for the failure; we need to fix it because most developers don't think in terms of `A` or `W` APIs -- the toolchain (.NET or C++ compilers) typically hide the complexity from you.

**Comment by AArnott** [comment]
Thanks, @ptorr-msft. Will you file a bug elsewhere to track accepting the non-suffixed method names in the future?
I've re-opened an issue on the PInvoke project to track adding the `W` suffix to all the applicable APIs in the meantime as a workaround. Thank you.

**Comment by kinex** [comment]
I am using only basic AES encryption APIs so my quick workaround was to replace PCLCrypto with .NET Standard APIs. This required adding a new .NET Standard lib for the encryption utilities into my solution. Only problem was that I need to call those encryption utilities from PCLs but a PCL cannot reference .NET Standard library and I am not ready to porting all my PCLs to .NET Standard yet. I solved this by using service locator pattern.

**Comment by mattrichnz** [comment]
@AArnott Do you have any update from the Microsoft folks? Unfortunately as users of PCLCrypto this is currently blocking us from submitting an update on the Windows store. We are using Win 10/UWP.

Many thanks!

**Comment by apwillies** [comment]
@AArnott We have this issue too, and the Windows ACK passes on our build server and local machines. It only fails when it gets to the Store. And also strange is that is passed on our Alpha and Beta flights on the store. It only fails when going out for final release. Any updates will be appreciated.

**Comment by HelenMamalaki** [comment]
This issue is for "Win8.x Store targeted apps" being rejected by the store, UWP apps with the latest Win10 SDK apps aren't rejected by the store because of PCLCrypto so I'm a bit confused about the Win8.1 vs UWP voting comments.

**Comment by apwillies** [comment]
@HelenMamalaki This issue is now also affecting UWP apps. Our app does not target 8.1 at all - only UWP 14393 and later.

**Comment by AArnott** [comment]
Microsoft has acknowledged a temporary issue with UWP that is supposed to be resolved in the near future, and an issue with Win8.1 apps that may or may not ever be fixed based on demand. There is talk of app authors obtaining "waivers" so their apps can get through even given these bugs in the store and SDK certification process. I'll see if one of those folks can comment on this thread to give you further direction, and to review the [votes above](https://github.com/AArnott/PCLCrypto/issues/97#issuecomment-316214175).

**Comment by deckertron9000** [comment]
@AArnott is your above comment based on your direct correspondence with Microsoft or based off of a discussion in forums/other internet communications regarding these issues? I'm curious if there's a link/page/issue tracker/forum post that talks more about the issue from Microsoft's side where we can weigh in as developers and possibly put some more pressure on Microsoft to fix this.

Or barring all that, do you know Microsoft's definition of "near future"? :joy:

**Comment by shawnharvell** [comment]
I'm dealing with this very issue right now.  I tried downgrading my PCLCrypto to a 1.x (1.0.86 to be precise) version as another poster suggested but with that I frequently get an exception when trying to decrypt.

Does anyone have a suggestion for a resolution?

And another question, the exception I get with 1.0.86 is "length does not a multiple of block size and no padding is selected ".  1.0.86 was the latest 1.x version available on NuGet.  Does anyone know of an earlier 1.x version that doesn't get that exception?

**Comment by AArnott** [comment]
@deckertron9000:

> is your above comment based on your direct correspondence with Microsoft

Yes. I work for Microsoft as well so I can use internal mailing lists to reach them.

> do you know Microsoft's definition of "near future"?

Yes, but it isn't for me to announce their timelines. I've invited them to respond on this thread directly though. All I can say is if you need to publish your app before it's fixed, they have a waiver process by which your app can be approved anyway. I've asked them to provide info for that process. But if your 'cert failed' result gives you a way to contest the result, I suggest you start that as that would probably be how to get a waiver.

@shawnharvell: the exception you describe sounds like a dupe of a [couple other reports](https://github.com/AArnott/PCLCrypto/issues?utf8=%E2%9C%93&q=is%3Aissue%20%22length%20does%20not%20a%20multiple%20%22) that never got resolved due to lack of response. It's quite possible that if it is indeed a library bug that 2.x fixed it during the rewrite to use BCrypt/NCrypt APIs.

**Comment by almudy** [comment]
Any news here? We ran in the exact same issue and are unable to submit our refreshed app to the store. Our paying customers are waiting. This cannot be true!

**Comment by AArnott** [comment]
@almudy Please vote up the appropriate [message above](https://github.com/AArnott/PCLCrypto/issues/97#issuecomment-316214175) based on your use case.

**Comment by AArnott** [comment]
Also in the meantime, @almudy, please check the store rejection result for a message about applying for a waiver.

**Comment by almudy** [comment]
I voted and will check the reject result soon. I just can't believe this is not fixed yet. This would never happen on Apple AppStore or Playstore...

**Comment by almudy** [comment]
Where exactly do I have to look for that message about applying for a waiver?

**Comment by AArnott** [comment]
Great question, @almudy. I've asked the Windows Store folks that question twice so I could provide it here. I'm still waiting on an answer from them.

**Comment by ptorr-msft** [comment]
There should be "notes to reviewers" when you submit the app. If you mention in there that you will fail WACK for `FindFirstFileEx` but that it is a known problem then hopefully they will give you the waiver. (I don't work for the Store team either, so can't guarantee anything). We are working on the fix but I can't give out dates for when it will be done, sorry.

**Comment by shawnharvell** [comment]
For the interested, I resolved my own issue by just removing PCLCrypto and going with an entirely different approach using another NuGet package that didn't have the same issue.  The way I was using PCLCrypto was pretty limited anyway so for me it wasn't a big deal.

**Comment by almudy** [comment]
My "notes to reviewers" are ignored. I think I have to find a way not to user PCLCrypto in our project.

**Comment by almudy** [comment]
I tried to use BouncyCastle instead of PCLCrypto. The implementation works but I still get the error when submitting to the store. 

Do I have to delete the following Nuget packages?

PCLCrypto
PInvoke.BCrypt
PInvoke.Kernel32
PInvoke.NCrypt
PInvoke.Windows.Core

Can somone give a hint how to check what exactly calls the problematic "FindFirstFileEx"?

Thanks in advance.

**Comment by almudy** [comment]
Just a word to clarify. I would clearly prefer to continue using PCLCrypto. But we do need a quick solution for our customers here.
Is it predictable if and at what time a updated Pinvoke.Kernel32 package will be available?

**Comment by apwillies** [comment]
@ptorr-msft  I already tried a few days ago to put in some "Notes for Certification" - the app gets rejected by an automated process and never gets to a human. I also used the "Certification Report Feedback" link twice - there was no response to this so far. 

I'm going to try to find a way to contact the Store with a kind of "Incident" today...

**Comment by kinex** [comment]
@almudy Yes those are the packages you have to uninstall (also the package "Validation" can be uninstalled). Do not delete packages directly from VS Solution explorer but use the Nuget package manager to make sure everything is uninstalled properly. As a final check you can make also a full text search into your project folders to make sure there are no references to "Kernel32" (that package is the main problem).

I recommend forgetting all crypto PCLs if you are developing for UWP or Xamarin, you don't need them and PCLs are starting to be deprecated stuff anyway. NET Standard libraries are the future and .NET Standard APIs already provide very good encryption APIs. Just create a new .NET Standard library project (for example .NETStandard 1.4) and try APIs from namespace System.Security.Cryptography.

**Comment by mattrichnz** [comment]
@AArnott @ptorr-msft This is a real issue for us right now too and is blocking an important release.
It is not like we even have the option of reverting to a previous version of PCLCrypto, as this too will be rejected by the certification process when submitting to the Windows Store.

Given adding "Notes for Certification" has no effect, this doesn't really leave us with too many options. Where is this recent change documented by Microsoft? If they plan to deprecate support for a particular native API, do they plan on advising customers and providing alternative recommendations?

**Comment by AArnott** [comment]
> It is not like we even have the option of reverting to a previous version of PCLCrypto, as this too will be rejected by the certification process when submitting to the Windows Store.

@mattrichnz why do you say that? PCLCrypto 1.x does *not* have this same problem and I fully expect it to be accepted by the Store. If you go this route, you should be sure that you're not still referencing the `PInvoke.*` packages, since those are the packages with the p/invoke's that the store cert process currently rejects.

> .NET Standard APIs already provide very good encryption APIs

Yes, @kinex's point is quite true. PCLCrypto is meant to help bridge the old world with itself and with newer libraries and apps. But if you're already "all new", in targeting platforms that have common crypto APIs available like newer .NETStandard versions, there's no need for PCLCrypto. Well *I take that back, partially*. The asymmetric encryption and key creation in Xamarin is *dog slow* (because it's software based). I'm talking minutes versus seconds here to derive a 4096 bit RSA key. PCLCrypto uses native APIs on all platforms, and those APIs leverage hardware acceleration, but Xamarin does not. So if you use RSA, there's a very compelling reason to stick with PCLCrypto anyway.

> Is it predictable if and at what time a updated Pinvoke.Kernel32 package will be available?

@almudy, PInvoke.Kernel32 gets updated regularly. But there isn't a change to be made to it to fix this problem. The library is not defective. The defect is in the Store certification process. What's more, some of the APIs the Store is rejecting are actually required by PCLCrypto 2.x, so we can't simply remove them either.

> Where is this recent change documented by Microsoft?

@mattrichnz I'm not sure what recent change you're talking about. But the plan as I understand it to rectify this situation is that the Store cert kit will be fixed to honor the MSDN documentation, thus permitting PCLCrypto and PInvoke libraries into the Store.

**Comment by almudy** [comment]
@AArnott Thanks for your detailed reply! So what do you think. Is this a question of days or weeks until MS will fix the Store certification process?

**Comment by HelenMamalaki** [comment]
Our UWP app is now also failing WACK because of the FindFirstFileEx API, however I've been told by a MS Support agent that a fix should be available by the end of this week. I will try publishing again next Monday 07/08/2017 and update.

**Comment by AArnott** [comment]
@almudy I can't say without approval from the Store folks. But @HelenMamalaki's report sounds very encouraging, I'd say.

**Comment by ptorr-msft** [comment]
@mattrichnz `FindFirstFileEx` has **not** been deprecated; please don't worry about that. This is a simple configuration error in the Store that is taking longer than expected to rectify.

(For future reference: if something *is* deprecated, we have a process for doing so that ensures things don't just disappear overnight. They are marked `deprecated` in the headers / metadata / MSDN for several releases before they're even *candidates* for removal, but even then we very rarely actually remove APIs from Windows).

**Comment by mattrichnz** [comment]
@ptorr-msft @AArnott Thanks for your answers, good to clarify a few points.

Fortunately, we've been able to switch over to .NET Standard libraries using System.Security.Cryptography and remove PCLCrypto. It is a shame as it was a great library but no doubt will use it again in the future once this issue is resolved.

**Comment by apwillies** [comment]
I opened a Support Incident today with the Store folks. To do this, go to the Store Dashboard, tap the question mark icon near the top right of the screen, and select Support. Then select Account Management and Setup and click either Chat Now or Submit an incident. I had a good response from the MS representative - helpful and patient, and acknowledged the issue is their side and hope to have it resolved by the end of the week. 

We also plan to switch to using the .NET Standard Crypto library - we'll need to check the speed issues mentioned by @AArnott. Also we still need to get our current submission out to production - hopefully that goes through by the end of the week using PCLCrypto.

**Comment by almudy** [comment]
Any news on this? My submission still gets rejected.

**Comment by AArnott** [comment]
I hear from the Store folks that a rollout with the fix was delayed when their testing discovered a bug in the new version. They're working to fix it.

**Comment by apwillies** [comment]
This is the current word from the Store folks, as at 8/8/17: 

> I can inform you that the issue has not been resolved and we do not have a current eta.

**Comment by apwillies** [comment]
The Store folks reached out to me today to resubmit our app. It passed and is busy publishing since about 2:30 p.m. EST 8/14/2017.

**Comment by AArnott** [comment]
@apwillies from your votes above it looks like yours is a UWP app that passed certification. That's great to hear.

**Comment by apwillies** [comment]
@AArnott Yes - it is up and running. Thanks for the help on this. I'm sure all the other interested parties will be happy also.

**Comment by almudy** [comment]
I can confirm this. Our app passed certification, too. Nice!

**Comment by AArnott** [comment]
I believe all UWP apps are unblocked now.
Win8 apps (which had a couple votes [up above](https://github.com/AArnott/PCLCrypto/issues/97#issuecomment-316214175)) may still experience a submission failure. That's still being worked on by the Store folks.

**Comment by Magic73** [comment]
So now the app pass the certification without any issue?

**Comment by almudy** [comment]
Yes

Von: Magic73 [mailto:notifications@github.com]
Gesendet: Dienstag, 12. September 2017 17:43
An: AArnott/PCLCrypto <PCLCrypto@noreply.github.com>
Cc: Alexander Handtke <alex@algrande.net>; Mention <mention@noreply.github.com>
Betreff: Re: [AArnott/PCLCrypto] Win8.x Store targeted apps are rejected due to Win32 crypto APIs (#97)


So now the app pass the certification without any issue?

**Comment by AArnott** [comment]
AFAIK Win8 store apps may still have an issue, since the last update I heard was that only UWP app submission had been fixed.
Can anyone confirm/deny that a Win8 app was accepted through Store validation?

**Comment by ItsRobbAllen** [comment]
Hello. We're targeting 8.1 apps and the latest cert kit fails it for the original issue above.
![pasted_image_at_2017_10_16_01_27_pm](https://user-images.githubusercontent.com/32247592/31673414-23abc100-b32d-11e7-95db-5776e4c8cecc.png)

**Comment by ItsRobbAllen** [comment]
The only thing we needed was MD5 hashing, so by removing the PInvoke libraries, we were able to pass the local cert kit.

**Comment by AArnott** [comment]
Thanks, @ItsRobbAllen. I've pinged the Windows Store certification folks to see if I can get the ball rolling for fixing 8.1 certification.

**Comment by AArnott** [comment]
@ItsRobbAllen Was this a local cert kit check that failed? If so, can you please download the latest SDK / cert check kit? The Store folks suggest that the latest one should pass (when run locally) and they expect the Store should not have rejected this.

**Comment by ozzy1873** [comment]
My app also failed certification. I am using PInvoke for encryption in a netstandard UWP app and would pull out the library entirely because I can now use the built in netstd functions, but unfortunately Microsoft.Azure.Mobile.Client 4.0.2 also has a dependency on PCLCrypto:
![image](https://user-images.githubusercontent.com/8699636/31773882-5ede077c-b4a1-11e7-94b6-cb72b6e4aac4.png)

**Comment by AArnott** [comment]
@ozzy1873 the failures you're seeing are due to a legit bug in PInvoke. It's defining the functions you've listed in the Store version of the library although those functions aren't available to Store apps. I will see those get fixed ASAP and you can update the PInvoke package used in your project by installing the latest one to apply the fix to your app. 
Is your screenshot the exhaustive list of errors, or are there more functions that fail the check?

**Comment by ozzy1873** [comment]
Yes, that is the exhaustive list of failures. Thanks for the quick turnaround!

Sent from Mail for Windows 10

**Comment by AArnott** [comment]
I'm closing this issue now that the Store issues have been resolved (with last remaining PInvoke issue tracked by the above issue).

**Comment by ItsRobbAllen** [comment]
I tried submitting to the store, but it failed for the same reason. I did update all the libraries as well as ensured I had the latest SDKs, but to no avail.

I put in the request ticket to MS this thread, I hope they can either give a waiver or at least inform me how to get our app to pass. The only thing we're using the library for is MD5 hashing as the Windows cryptography library won't work on the iPad, and both our apps share the same code bases.

**Comment by ozzy1873** [comment]
@ItsRobbAllen I believe this has been fixed, just not yet published.

**Comment by AArnott** [comment]
@ozzy1873 Are you talking about the PInvoke issue you called out before? Because @ItsRobbAllen hit a different set of errors that suggest to me that it's a store cert problem, as opposed to the list you (@ozzy1873) hit that was due to a set of methods that PInvoke truly shouldn't have had for Store apps.

**Comment by ozzy1873** [comment]
Yes, I was talking about the PInvoke issue I called out before. Sorry, when he said it failed for the same reason, I assumed it was the same set of errors.
